package/src/cpumon.cpp

```diff
@@ -63,60 +63,62 @@ std::map<std::string, double> const cpumon::get_json_average_stats(
-
-  // Loop over the file
-  std::string line;
-  while (std::getline(cpuInfoFile, line)) {
+  // Map lscpu names to the desired ones in the JSON
+  const std::map<std::string, std::string> metricToName{
+      {"Model name", "ModelName"},
+      {"CPU(s)", "CPUs"},
+      {"Socket(s)", "Sockets"},
+      {"Core(s) per socket", "CoresPerSocket"},
+      {"Thread(s) per core", "ThreadsPerCore"}};
+
+  // Useful function to determine if a string is purely a number
+  auto isNumber = [](const std::string& s) {
+    return !s.empty() && std::find_if(s.begin(), s.end(), [](unsigned char c) {
+                           return !std::isdigit(c);
+                         }) == s.end();
+  };
+
+  // Loop over the output, parse the line, check the key and store the value if
+  // requested
+  std::string key{}, value{};
+
+  for (const auto& line : cmd_result.second) {
+    // Continue on empty line
     if (line.empty()) continue;
+
+    // Tokenize by ":"
     size_t splitIdx = line.find(":");
-    std::string val;
-    if (splitIdx != std::string::npos) {
-      val = line.substr(splitIdx + 1);
-      if (val.empty()) continue;
-      for (const auto& metric : metrics) {
-        if (line.size() >= metric.size() &&
-            line.compare(0, metric.size(), metric) == 0) {
-          if (metric == "processor")
-            nCPU++;
-          else if (metric == "siblings") {
-            if (nSiblings == 0) nSiblings = std::stoi(val);
-          } else if (metric == "cpu cores") {
-            if (nCores == 0) nCores = std::stoi(val);
-          } else if (hw_json["HW"]["cpu"][metric].empty()) {
-            hw_json["HW"]["cpu"][metric] =
-                std::regex_replace(val, std::regex("^\\s+|\\s+$"), "");
-          }
-        }  // end of metric check
-      }    // end of populating metrics
-    }      // end of seperator check
-  }        // end of reading cpuInfoFile
-
-  // The clear assumption here is that there is a single type of processor
-  // installed nCPU = nSockets * nSiblings nSiblings = nCoresPerSocket *
-  // nThreadsPerCore
-  // Next 4 lines are a workaround for Raspberry Pi
-  if (!nSiblings) nSiblings = nCPU;
-  if (!nCores) nCores = nCPU;
-  unsigned int nSockets = nCPU / nSiblings;
-  unsigned int nThreads = nSiblings / nCores;
-  hw_json["HW"]["cpu"]["nCPU"] = nCPU;
-  hw_json["HW"]["cpu"]["nSockets"] = nSockets;
-  hw_json["HW"]["cpu"]["nCoresPerSocket"] = nCores;
-  hw_json["HW"]["cpu"]["nThreadsPerCore"] = nThreads;
-
-  // Close the file
-  cpuInfoFile.close();
+    if (splitIdx == std::string::npos) continue;
+
+    // Read "key":"value" pairs
+    key = line.substr(0, splitIdx);
+    value = line.substr(splitIdx + 1);
+    if (key.empty() || value.empty()) continue;
+    key = std::regex_replace(key, std::regex("^\\s+|\\s+$"), "");
+    value = std::regex_replace(value, std::regex("^\\s+|\\s+$"), "");
+
+    // Fill the JSON with the information
+    for (const auto& metric : metricToName) {
```

Comment:
Oh but wait... now we can use the map:

```cpp
if (metricToName.count(metric) == 1) {
...
}
```

That's the right way to do it!

Reply:
Good point, changed now. I seldom use maps, as a deliberate choice, so this completely escaped me. Is this still an `O(N)` operation or is it actually smarter than that?

Reply:
Wait I got it wrong 😄

Reply:
OK, we should be good now, today is not my day I guess.

Reply:
I think we are both ready for the weekend!

Reply:
> Good point, changed now. I seldom use maps, as a deliberate choice, so this completely escaped me. Is this still an `O(N)` operation or is it actually smarter than that?

As it's a map (with a hash) it's usually constant (with the worst case being linear).

Reply:
The main advantage here of using a map is it makes the program's intention clearer than looping over a vector, which means less chance of screwing up... 😁